addons/http_routing/models/ir_http.py

```diff
@@ -432,18 +432,20 @@ def _match(cls, path):
         # requires one to set the lang on the request. Temporary grant
         # the public user. Don't try it at home!
         real_env = request.env
-        try:
-            request.registry['ir.http']._auth_method_public()  # it calls update_env
-            nearest_url_lang = cls.get_nearest_lang(request.env['res.lang']._lang_get_code(url_lang_str))
-            cookie_lang = cls.get_nearest_lang(request.httprequest.cookies.get('frontend_lang'))
-            context_lang = cls.get_nearest_lang(real_env.context.get('lang'))
-            default_lang = cls._get_default_lang()
-            request.lang = request.env['res.lang']._lang_get(
-                nearest_url_lang or cookie_lang or context_lang or default_lang._get_cached('code')
-            )
-            request_url_code = request.lang._get_cached('url_code')
-        finally:
-            request.env = real_env
+        # try:
+        request.registry['ir.http']._auth_method_public()  # it calls update_env
+        nearest_url_lang = request.env['res.lang']._lang_get_code(url_lang_str)                # url_code from URL
+        cookie_lang = request.env['res.lang']._lang_get_code(request.httprequest.cookies.get('frontend_lang')) # url_code from cookie, returning visitor
+        context_lang = cls.get_nearest_lang(real_env.context.get('lang'))                                            # code from browser
+        default_lang = cls._get_default_lang()                                                 # string from browser
+
+        # pylint: disable=assigning-non-slot
+        request.lang = request.env['res.lang']._lang_get(
+            nearest_url_lang or cookie_lang or context_lang or default_lang._get_cached('code')
+        )
+        request_url_code = request.lang._get_cached('url_code')
+        # finally:
+        #     request.env = real_env
```

Comment:
I can put back the `try: finally:` but I'll check if we really need to change the environment of the request? if we can avoid, it's better.  Do you know if there is a test about that? (because this new implementation is green on runbot without it).

> The comment about the lang from the context is wrong, it is the lang of the connected user
My understanding is that it can be a mix; lang of the user, or default lang of the browser; for example in `odoo/http.py`, it sets `context['lang']` to the brower lang. It's the reason I could not remove `get_nearest_lang`, but the goal would be to clean context and remove the `get_nearest_lang`.

Reply:
About the try/finally, I don't think it's really needed, it was defensing programming and can probably be removed: https://github.com/odoo/odoo/pull/78857#discussion_r755301678

Reply:
> but I'll check if we really need to change the environment of the request?

There can be no user yet on the environment (in case of a visitor) but accessing `res.lang` requires querying the database hence we (temporary) grant the public user for those queries. The try/finally is about restoring the original user (maybe `None`, no user) once the queries on `res.lang` are done. In case we remove the try/finally and that there is an exception inside those lines, the environment would still have the public user, it would had leaked. This try/finally is just a common cleanup pattern, it's basically like writing:

```py
with public_user:
    ...
```

But there is no such thing as `with public_user:`, hence the try/finally.

I don't get why you would want to remove that... Ok there is no error atm but then what? A try/finally is basically a no-op when no error occurred. So it's basically free and it is more defensive in case an error actually occurs.

Just... read the comment above "don't try this at home" and trust the fw-py guy please...

> My understanding is that it can be a mix; lang of the user, or default lang of the browser; for example in odoo/http.py, it sets context['lang'] to the brower lang. It's the reason I could not remove get_nearest_lang, but the goal would be to clean context and remove the get_nearest_lang.

You are right, when the user is not logged in (visitor) then we use the lang of the brower to ensure that there always is a lang set in the context. That's an exception from all the other context values that are exclusively taken from the session context, which in turn are set according to the preference set on the user's `res.user` record. So I think it is better to document it that way: lang from the session.

Also I notice it now, `cls._get_default_lang()` uses `tools.get_lang()` which doesn't get the lang from the browser. That comment should be fixed too.

Reply:
Edit, I didn't see that the comments were fixed. It is fine :)